File: couleurs.py
import numpy as np
import cv2 as cv

ComputerCamera=0;
Black=(0,0,0)

TextFont=cv.FONT_HERSHEY_SIMPLEX

# Blue
BlueMin = np.array([94, 80, 2], np.uint8)
BlueMax = np.array([120, 255, 255], np.uint8)

# Red
RedMin = np.array([136, 87, 111], np.uint8)
RedMax = np.array([180, 255, 255], np.uint8)

# Green
GreenMin = np.array([25, 52, 72], np.uint8)
GreenMax = np.array([102, 255, 255], np.uint8)

cap=cv.VideoCapture(ComputerCamera)

if not cap.isOpened():
    print("Camera cannot be opened")
    exit()
while True :
    ret, frame=cap.read()

    if not ret:
        print("Error access to camera data")
        break

    BgrToHsvFrame=cv.cvtColor(frame,cv.COLOR_BGR2HSV)

    #BgrToGrayFrame = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
    #BgrToGrayFrame = np.float32(BgrToGrayFrame)

    #CornerHarris=cv.cornerHarris(BgrToGrayFrame,2,3,0.04)
    #CornerHarris=cv.dilate(CornerHarris,None)

    #frame[CornerHarris>0.01*CornerHarris.max()]=[0,0,255]

    BlueMask = cv.inRange(BgrToHsvFrame, BlueMin, BlueMax)
    RedMask = cv.inRange(BgrToHsvFrame, RedMin, RedMax)
    GreenMask = cv.inRange(BgrToHsvFrame, GreenMin, GreenMax)

    kernel = np.ones((5, 5), "uint8")

    BlueMask = cv.dilate(BlueMask, kernel)
    res_blue = cv.bitwise_and(frame, frame,
                               mask=BlueMask)

    RedMask = cv.dilate(RedMask, kernel)
    res_red = cv.bitwise_and(frame, frame,
                              mask=RedMask)

    GreenMask = cv.dilate(GreenMask, kernel)
    res_green = cv.bitwise_and(frame, frame,
                             mask=GreenMask)

    # Creating contour to track blue color
    contours, hierarchy = cv.findContours(BlueMask,
                                           cv.RETR_TREE,
                                           cv.CHAIN_APPROX_SIMPLE)
    for pic, contour in enumerate(contours):
        area = cv.contourArea(contour)
        if (area > 300):
            x, y, w, h = cv.boundingRect(contour)
            imageFrame = cv.rectangle(frame, (x, y),
                                       (x + w, y + h),
                                       (255, 0, 0), 2)

            cv.putText(frame, "Bleu", (x, y),
                        TextFont,
                        1.0, (255, 0, 0))

    contours, hierarchy = cv.findContours(RedMask,
                                           cv.RETR_TREE,
                                           cv.CHAIN_APPROX_SIMPLE)

    # Creating contour to track red color
    for pic, contour in enumerate(contours):
        area = cv.contourArea(contour)
        if (area > 300):
            x, y, w, h = cv.boundingRect(contour)
            imageFrame = cv.rectangle(frame, (x, y),
                                       (x + w, y + h),
                                       (0, 0, 255), 2)

            cv.putText(imageFrame, "Rouge", (x, y),
                        cv.FONT_HERSHEY_SIMPLEX, 1.0,
                        (0, 0, 255))

    # Creating contour to track green color
    contours, hierarchy = cv.findContours(GreenMask,
                                           cv.RETR_TREE,
                                           cv.CHAIN_APPROX_SIMPLE)

    for pic, contour in enumerate(contours):
        area = cv.contourArea(contour)
        if (area > 300):
            x, y, w, h = cv.boundingRect(contour)
            imageFrame = cv.rectangle(frame, (x, y),
                                       (x + w, y + h),
                                       (0, 255, 0), 2)

            cv.putText(imageFrame, "Vert", (x, y),
                        cv.FONT_HERSHEY_SIMPLEX,
                        1.0, (0, 255, 0))

    #gray=cv.cvtColor(frame,cv.COLOR_BGR2GRAY)
    #cv.imshow('frame')

    #cv.rectangle(frame,RectangleTopLeftCorner,RectangleBotRightCorner,Black,2)
    #cv.putText(frame,"Testing...",(10,180),TextFont,4,Black,2)

    cv.imshow('frame',frame)

    if cv.waitKey(1)==ord('q'):
        break

cap.release()
cv.destroyAllWindows()

File: couleurs.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <string>

int main() {
    cv::VideoCapture cap(0); // Utilisation de la première caméra

    if (!cap.isOpened()) {
        std::cerr << "Erreur : la caméra ne peut pas être ouverte." << std::endl;
        return -1;
    }

    cv::Mat frame, hsvFrame;
    cv::namedWindow("frame", cv::WINDOW_AUTOSIZE);

    // Définition des plages de couleurs HSV pour le filtre
    cv::Scalar blueMin(94, 80, 2), blueMax(120, 255, 255);
    cv::Scalar redMin(136, 87, 111), redMax(180, 255, 255);
    cv::Scalar greenMin(25, 52, 72), greenMax(102, 255, 255);
    cv::Scalar whiteMin(0, 0, 200), whiteMax(180, 30, 255);
    cv::Scalar blackMin(0, 0, 0), blackMax(180, 255, 30);

    std::vector<std::pair<cv::Scalar, cv::Scalar>> colorRanges = {
        {blueMin, blueMax}, {redMin, redMax}, {greenMin, greenMax},
        {whiteMin, whiteMax}, {blackMin, blackMax}
    };

    std::vector<std::string> colorNames = {"Bleu", "Rouge", "Vert", "Blanc", "Noir"};
    std::vector<cv::Scalar> drawColors = {
        cv::Scalar(255, 0, 0), cv::Scalar(0, 0, 255), cv::Scalar(0, 255, 0),
        cv::Scalar(255, 255, 255), cv::Scalar(0, 0, 0)
    };

    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(5, 5));

    while (true) {
        bool ret = cap.read(frame);
        if (!ret) {
            std::cerr << "Erreur : impossible d'accéder aux données de la caméra." << std::endl;
            break;
        }

        cv::cvtColor(frame, hsvFrame, cv::COLOR_BGR2HSV);

        for (size_t i = 0; i < colorRanges.size(); ++i) {
            cv::Mat mask;
            cv::inRange(hsvFrame, colorRanges[i].first, colorRanges[i].second, mask);
            mask = cv::dilate(mask, kernel);

            std::vector<std::vector<cv::Point>> contours;
            cv::findContours(mask, contours, cv::RETR_TREE, cv::CHAIN_APPROX_SIMPLE);

            for (size_t j = 0; j < contours.size(); ++j) {
                double area = cv::contourArea(contours[j]);
                if (area > 300) {
                    cv::Rect boundingBox = cv::boundingRect(contours[j]);
                    cv::rectangle(frame, boundingBox, drawColors[i], 2);
                    cv::putText(frame, colorNames[i], cv::Point(boundingBox.x, boundingBox.y),
                                cv::FONT_HERSHEY_SIMPLEX, 1.0, drawColors[i]);
                }
            }
        }

        cv::imshow("frame", frame);

        if (cv::waitKey(1) == 'q') {
            break;
        }
    }

    cap.release();
    cv::destroyAllWindows();
    return 0;
}
